src/parameters/mod.ts

```diff
@@ -21,46 +37,54 @@ export type PossibleParameterKeys<
 > = (keyof ParameterSetInternal)[];
 
 export type ParameterPropertiesDefinition<
-  Parameters extends ParameterSetDefinition,
-  Required extends PossibleParameterKeys<Parameters>,
+  Params extends ParameterSetDefinition,
+  Required extends PossibleParameterKeys<Params>,
 > = {
-  properties: Parameters;
+  properties: Params;
   required: Required;
 };
 
 export type ParameterVariableType<
   Def extends ParameterDefinition,
-> = Def["type"] extends ICustomType // If the ParameterVariable is a Custom type, use it's definition instead
-  ? ParameterVariableType<Def["type"]["definition"]>
-  : Def extends TypedObjectParameterDefinition // If the ParameterVariable is of type object, allow access to the object's properties
-    ? ObjectParameterVariableType<Def>
+  CurrentDepth extends RecursionDepthLevel = 0,
+> = CurrentDepth extends MaxRecursionDepth ? UntypedObjectParameterVariableType // i.e. any
+  : Def extends CustomTypeParameterDefinition // If the ParameterVariable is a Custom type, use it's definition instead
+    ? ParameterVariableType<
+      Def["type"]["definition"],
+      IncreaseDepth<CurrentDepth>
+    >
+  // If the ParameterVariable is of type object, allow access to the object's properties
+  : Def extends TypedObjectParameter ? ObjectParameterVariableType<Def>
   : Def extends UntypedObjectParameterDefinition
     ? UntypedObjectParameterVariableType
   : SingleParameterVariable;
 
 // deno-lint-ignore ban-types
-type SingleParameterVariable = {};
+export type SingleParameterVariable = {};
 
 // deno-lint-ignore no-explicit-any
 type UntypedObjectParameterVariableType = any;
 
-type ObjectParameterPropertyTypes<Def extends TypedObjectParameterDefinition> =
-  {
-    [name in keyof Def["properties"]]: ParameterVariableType<
-      Def["properties"][name]
-    >;
-  };
+export type ObjectParameterPropertyTypes<
+  Props extends TypedObjectProperties,
+> = {
+  [name in keyof Props]: ParameterVariableType<
+    Props[name]
+  >;
+};
 
 // If additionalProperties is set to true, allow access to any key.
 // Otherwise, only allow keys provided through use of properties
-type ObjectParameterVariableType<Def extends TypedObjectParameterDefinition> =
-  Def["additionalProperties"] extends false ? ObjectParameterPropertyTypes<Def>
-    : 
-      & ObjectParameterPropertyTypes<Def>
-      & {
-        // deno-lint-ignore no-explicit-any
-        [key: string]: any;
-      };
+type ObjectParameterVariableType<
+  Def extends TypedObjectParameter,
+> = Def["additionalProperties"] extends false
+  ? ObjectParameterPropertyTypes<Def["properties"]>
+  : 
+    & ObjectParameterPropertyTypes<Def["properties"]>
+    & {
+      // deno-lint-ignore no-explicit-any
+      [key: string]: any;
+    };
```

Comment:
Nit: We could do what we did for runtime functions and remove the need to specify `ObjectParameterPropertyTypes<Def["properties"]>` twice

```suggestion
type ObjectParameterVariableType<
  Def extends TypedObjectParameter,
> =
  & ObjectParameterPropertyTypes<Def["properties"]>
  & (Def["additionalProperties"] extends false ? Record<never, never>
    : {
      // deno-lint-ignore no-explicit-any
      [key: string]: any;
    });
```